WikipediaUnitTests/Code/LocationManagerTests.swift

```
@@ -0,0 +1,230 @@
+        XCTAssertEqual(locationManager.location, location)
+    }
+
+    func testUpdateHeading() {
+        locationManager.startMonitoringLocation()
+
+        let heading = MockCLHeading(headingAccuracy: 10)
+        mockCLLocationManager.simulateUpdate(heading: heading)
+        XCTAssertEqual(locationManager.heading, heading) // Test new heading on location manager
+        XCTAssertEqual(delegate.heading, heading)  // Test new heading on delegate
+    }
+
+    // When the heading has already been fetched, a new instance of CLLocationManager contains the last known heading in its heading variable, even before startMonitoringLocation() has been called
+    func testExistingHeadingWithoutCallback() {
+        let heading = MockCLHeading(headingAccuracy: 10)
+        mockCLLocationManager.simulateUpdate(heading: heading)
+        XCTAssertEqual(locationManager.heading, heading) // Test last know heading on location manager
+    }
+
+    func testStopUpdating() {
+        locationManager.startMonitoringLocation()
+        // Setup value when monitoring
+        let location1 = CLLocation(latitude: 10, longitude: 20)
+        mockCLLocationManager.simulateUpdate(location: location1)
+        let heading1 = MockCLHeading(headingAccuracy: 10)
+        mockCLLocationManager.simulateUpdate(heading: heading1)
+
+        locationManager.stopMonitoringLocation()
+        // Update with new values after stop monitoring for location
+        let location2 = CLLocation(latitude: 100, longitude: 200)
+        mockCLLocationManager.simulateUpdate(location: location2)
+        let heading2 = MockCLHeading(headingAccuracy: 100)
+        mockCLLocationManager.simulateUpdate(heading: heading2)
+
+        // Heading and location
+        XCTAssertEqual(locationManager.heading, heading1)
+        XCTAssertEqual(delegate.heading, heading1)
+        XCTAssertEqual(locationManager.location, location1)
+        XCTAssertEqual(delegate.location, location1)
+
+        // Error
```

// Check the error is not propagated